### ISSUE TITLE
Multiview fusion in nuScenes

### ISSUE THREAD

**Issue by kwonyoung9120** [pengsongyou/openscene#23]
Thanks for the excellent work! 
I'm trying to do multi-view feature fusion in nuScenes dataset. In code, it seems that 6 images are fused for each scan. 
I wonder that the feature could be robust if I apply multi-view fusion with 'adjacent 30 images' (from adjacent 5 LiDAR scans).
When I preprocess the openseg feature as I proposed, it records 32.37 mIoU in nuScene validation set.  (Only used the 2D multiview features)

1. I wonder how much mIoU should be record applying with only 2D openseg features.  2D-3D ensemble records about 42, so it would be slightly lower I guess. (Maybe over than 35.xx..?)
2. Is there any ablation or opinion for number of the multi-view frames?

**Comment by kwonyoung9120** [comment]
Oh I inference this and recorder 34.03 mIoU with 2D fused feature only. Maybe using other images give noisy features than using just 6 images per scan. (due to dynamic objects or occlusions) Closing this issue.

**Comment by pengsongyou** [comment]
Hi @kwonyoung9120 

Which features did you use for inference to get 34.03 mIoU with 2D fused features? When did you download the preprocessed 3D point clouds and fused features? I am asking because I was evaluating again the 2D fused features that I provided roughly a month ago, but I could not really get what I got before (33-34 mIoU). Strangely, Just wanna figure out if anything that I changed recently in my code led to this strange drop :)

**Comment by kwonyoung9120** [comment]
Thanks for the kind reply! 

I just preprocess nuScene validation set with Openseg model(I just ran this one /scripts/feature_fusion/nuscenes_openseg.py with some slight changes to load directly from the nuScene meta-data by nusc-devkit). From this (which pool just 6 images per scan I guess) get 34.03 mIoU for today. I didn't download from your link because somehow it is too slow.. 

Oh I have another question about the nuScene dataset. In Tab.4 from the main paper, ScanNet and Matterport3D shows better results in 3D distillation than 2D fusion. Is this tendency shows same in nuScene dataset? I'm asking this because when I distill MinkNet18A directly, the mIoU shows similar with 2D fusion but not more than this(31.xx (3D) < 32.xx(2D)). (It might be from the learning parameter difference hmm..)

**Comment by pengsongyou** [comment]
Hi, thanks for the information!

Regarding your question: yes! I should have included table ablation for nuScenes as well. As you know, you could get ~34 mIoU from 2D fused feature. For the distillation model, you can simply try to use the provided pretrained model from me, and you should be able to get 42 mIoU already. The final ensemble model is also around 42 mIoU, but with higher mAcc (57 vs 61). Make sure that your input point cloud is the entire point cloud from a 0.5 sec sequence, but evaluate on the last timestamp. I belive this is also done in previous methods in nuScenes benchmark.

Best
Songyou

**Comment by pengsongyou** [comment]
> _Make sure that your input point cloud is the entire point cloud from a 0.5 sec sequence, but evaluate on the last timestamp_ -> So that means, I should preprocess about 5 scans(4 unlabeled sweeps and 1 scans with semantic annotation, not sure) with 6 images, not the single annotated scan. It seems that this multi-frame input setup get some gain to train the nuScenes dataset.
> 
So, each 0.5 sec sequence contains all lidar point clouds within this 0.5 seconds (I don't remember but should be 5-6 frames in total?). However, the semantic annotation is only done on the points from the last timestamp, and that is where we evaluate on. For the feature fusion, we also only use the synchronized images and the corresponding Lidar
points on those points from the last timestamp. 

Still, to train a MinkowskiNet, we need to input all points from the entire 0.5 seconds, but supervise only with the features on the last timestamp (this is to incorporate more information from frames nearby to help the prediction).

Hope that makes sense.

Best
Songyou

**Comment by kwonyoung9120** [comment]
> > _Make sure that your input point cloud is the entire point cloud from a 0.5 sec sequence, but evaluate on the last timestamp_ -> So that means, I should preprocess about 5 scans(4 unlabeled sweeps and 1 scans with semantic annotation, not sure) with 6 images, not the single annotated scan. It seems that this multi-frame input setup get some gain to train the nuScenes dataset.
> 
> So, each 0.5 sec sequence contains all lidar point clouds within this 0.5 seconds (I don't remember but should be 5-6 frames in total?). However, the semantic annotation is only done on the points from the last timestamp, and that is where we evaluate on. For the feature fusion, we also only use the synchronized images and the corresponding Lidar points on those points from the last timestamp.
> 
> Still, to train a MinkowskiNet, we need to input all points from the entire 0.5 seconds, but supervise only with the features on the last timestamp (this is to incorporate more information from frames nearby to help the prediction).
> 
> Hope that makes sense.
> 
> Best Songyou

That's a relief that I can extract Openseg features only for the annotated LiDAR. Thank you for the detail comments!

**Comment by Charlie839242** [comment]
> I just preprocess nuScene validation set with Openseg model(I just ran this one /scripts/feature_fusion/nuscenes_openseg.py with some slight changes to load directly from the 'original' nuScene with nusc-devkit). From this (which pool just 6 images per scan I guess) get 34.03 mIoU for today. I didn't download from your link because somehow it is too slow..

Hi, would mind sharing the modified script for generating fused features for nuScenes? I follow the guide and can only obtain result around 32.3. Thanks a lot!

**Comment by Charlie839242** [comment]
> Which features did you use for inference to get 34.03 mIoU with 2D fused features? When did you download the preprocessed 3D point clouds and fused features? I am asking because I was evaluating again the 2D fused features that I provided roughly a month ago, but I could not really get what I got before (33-34 mIoU). Strangely, Just wanna figure out if anything that I changed recently in my code led to this strange drop :)


Hi pengsongyou! I am wondering did you find the difference? I follow your code and can only obtain an miou of 32%  with 2d features only on nuScenes dataset. Looking forward to your reply!
